tools/shared/src/main/scala/org/scalajs/core/tools/linker/analyzer/Analyzer.scala

```diff
@@ -182,22 +191,29 @@ private final class Analyzer(semantics: Semantics,
 
     var nonExistent: Boolean = false
 
-    /** Ensures that this class and its dependencies are linked.
-     *
-     *  @throws CyclicDependencyException if this class is already linking
-     */
-    def linkClasses(): Unit = {
-      if (_linking)
-        throw CyclicDependencyException(encodedName :: Nil)
+    /** Ensures that this class and its dependencies are linked. */
+    def linkClasses(): Unit =
+      linkClasses(linkingFrom = null)
 
+    private def linkClasses(linkingFrom: ClassInfo): Unit = {
       if (!_linked) {
-        _linking = true
-        try {
+        if (_linkingFrom == null)
+          _linkingFrom = linkingFrom
+
+        if (!_linking) {
+          _linking = true
           linkClassesImpl()
-        } catch {
-          case CyclicDependencyException(chain) =>
-            throw CyclicDependencyException(encodedName :: chain)
+        } else {
+          def makeCycle(from: ClassInfo, acc: List[ClassInfo]): List[ClassInfo] = {
+            if (from == linkingFrom) from :: acc
+            else makeCycle(from._linkingFrom, from :: acc)
+          }
```

Comment:
That's probably easier. Optionally you could also have `linkClasses` an `Option[List[ClassInfo]]` so we don't have to use exceptions for control flow... But I think in this case its actually ok.